framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,691 @@
+        printWriter.print(formatAFieldImpl(aField, fieldName, /*enclosing class=*/ null));
+        printWriter.println(";");
+        printWriter.println();
+    }
+
+    /**
+     * Prints a method declaration in stub file format (i.e., without a method body).
+     *
+     * @param aMethodWrapper the method to print
+     * @param basename the simple name of the enclosing class, for receiver parameters and
+     *     constructor names
+     * @param printWriter where to print the method signature
+     * @param indentLevel the indent string
+     */
+    private static void printMethodDeclaration(
+            AMethodWrapper aMethodWrapper,
+            String basename,
+            PrintWriter printWriter,
+            String indentLevel) {
+
+        AMethod aMethod = aMethodWrapper.getAMethod();
+
+        printWriter.print(indentLevel);
+
+        printTypeParameters(aMethodWrapper.getTypeParameters(), printWriter);
+
+        String methodName = aMethod.getMethodName();
+        // Use Java syntax for constructors.
+        if ("<init>".equals(methodName)) {
+            methodName = basename;
+        } else {
+            printWriter.print(formatType(aMethod.returnType, aMethodWrapper.getReturnType()));
+        }
+        printWriter.print(methodName);
+        printWriter.print("(");
+
+        StringJoiner parameters = new StringJoiner(", ");
+
+        if (!aMethod.receiver.type.tlAnnotationsHere.isEmpty()) {
+            // Only output the receiver if it has an annotation.
+            parameters.add(
+                    formatParameter(
+                            AFieldWrapper.createReceiverParameter(aMethod.receiver),
+                            "this",
+                            basename));
+        }
+        for (Integer index : aMethodWrapper.getParameters().keySet()) {
+            AFieldWrapper param = aMethodWrapper.getParameters().get(index);
+            parameters.add(formatParameter(param, param.getParameterName(), basename));
+        }
+        printWriter.print(parameters.toString());
+        printWriter.println(");");
+        printWriter.println();
+    }
+
+    /**
+     * The implementation of {@link #write(ASceneWrapper, Writer)} and {@link #write(ASceneWrapper,
+     * String)}. Prints imports, classes, method signatures, and fields in stub file format, all
+     * with appropriate annotations.
+     *
+     * @param scene the scene to write
+     * @param printWriter where to write the scene to
+     */
+    private static void writeImpl(ASceneWrapper scene, PrintWriter printWriter) {
+        // Write out all imports
+        ImportDefWriter importDefWriter;
+        try {
+            importDefWriter = new ImportDefWriter(scene, printWriter);
+        } catch (DefException e) {
+            throw new BugInCF(e);
+        }
+        importDefWriter.visit();
+        printWriter.println();
+
+        // sort by package name so that output is deterministic and default package
```

Comment:
For determinism, it is necessary to sort by the fully-qualified class name, or else the classes within a package may be output in nondeterministic order.  (Unless they were already sorted and `Collections.sort` is a stable sort, in which case document those facts.)